### ISSUE TITLE
ACL AUTH support breaks the legacy auth format

### ISSUE THREAD

**Issue by fgm** [gomodule/redigo#561]
The DSN format used since #480 supports the general https://www.iana.org/assignments/uri-schemes/prov/redis format.

The actual DSN formats used by `redis-cli` (and possibly other tools) are:

- `redis://user:@`  : "user" user, empty password, ACL mode
- `redis://:pass@`: "empty user, "pass" password, ACL mode
- `redis://pass@` : requirepass mode

However, with redigo &ge; 1.8.4 the behavior differs depending on whether the connection is to DB#0 or any other. 

- `redis://user:@/0`  : "user" user, empty password, ACL mode
- `redis://:pass@/0` : "empty user, "pass" password, ACL mode
- `redis://pass@/0` : requirepass mode -> succeeds
- `redis://user:@/1`  : "user" user, empty password, ACL mode
- `redis://:pass@/1` : "empty user, "pass" password, ACL mode
- `redis://pass@/1` : requirepass mode -> fails with _failed dialing Redis URL: NOAUTH Authentication required._

What is happening is that in the `redis://<pass>@/<db>` format, without the `":"` in the user part of the DSN, the password is parsed as being `u.User.username` instead of `u.User.password`. This causes the `if isSet` in `DialURL` to be skipped, as the password is being returned as the user. 

At this point, the behavior differs for DB#0 : unlike what happens for the other DBs, the connection is returned, correctly open and unauthenticated. It can still be authenticated manually by the app. For other DBs, however, Redigo emits a
`SELECT` call, which fails for lack of authentication.

For consistency and compatibility with legacy and redis-cli behavior, that mode should also be supported for non-0 DBs.

**Comment by stevenh** [comment]
The userinfo in the [RFC](https://www.iana.org/assignments/uri-schemes/prov/redis) requires the `:`
> The "userinfo" field uses the traditional "user:password" format.

It also specifically advises against using the username for anything else
> Until/unless that happens: URI producers are advised to
  leave the username portion of the "userinfo" URI field blank; URI
  consumers are advised to be aware of the future possibility of
  non-blank username portions of URIs; attempting to use the username
  portion of URIs for any purpose other than specifying the username to
  use when authenticating to the Redis server is strongly advised
  against.

In your example `redis://<password>@/<db>` the format is invalid as it's missing the `:` if we we're to assume a more flexible format the parameter you refer to as `<password>` would actually be `<user>`.

I've played with a few examples, see below, to check if db 0 and db 1 exhibit different behaviours and they appear consistent
`redis://:abc123@/0`
`redis://:abc123@/1`

Looking at the changes in [PR](https://github.com/gomodule/redigo/pull/480) I've been unable to identify any inconsistent behaviour as the check was always using the values from `u.User.Password()` which given `redis://abc123@/0` would result in a isSet always being `false`.

**Comment by fgm** [comment]
The change happens as described when not using the ":", specifically for the requirepass mode with `redis://pass@....`.

The behaviour described is what current redis-cli does, which does indeed not completely match the RFC, but is as authoritative an implementation as one can imagine. And the crux of the matter is how differently the package behaves in this "traditional" format between DB0 and the other DBs, while redis-cli remains consistent. As described, the change is the new ` if isSet` in DialURL, leading to these different behaviours.

**Comment by stevenh** [comment]
Sorry @fgm I'm still not clear, could you provide an example that worked before the PR I've tried the following and they both fail.
`redis://abc123@/0`
`redis://abc123@/1`

I'm also confused by:
> the change is the new if isSet in DialURL

As that change has nothing to do with #480 and is over 5 years old.

**Comment by fgm** [comment]
Sorry, my original post was probably not clear enough: I'm not saying the issue stems from PR #480, only pointing to that PR as the place where both formats started to be supported. It's actually not really relevant and led you astray. The issue has actually been present since https://github.com/gomodule/redigo/commit/5e9d3af5852c9ba2b6c911b5b990b2398d70cf27 introduced AUTH to fix https://github.com/gomodule/redigo/issues/134 .

Here is a small program demonstrating the problem:

```go
package main

import (
	"log"

	"github.com/gomodule/redigo/redis"
)

const pass = "legacypass"

func ping(conn redis.Conn) {
	log.Println(conn.Do("AUTH", pass))
	log.Println(conn.Do("PING"))
}

func main() {
	var c redis.Conn
	var err error

	// Before running this, configure a redis.with "requirepass" security on port 6379, no ACL.
	c, err = redis.DialURL("redis:///0")
	if err != nil {
		log.Printf("DB0: %v", err)
	} else {
		log.Println("DB0: got connection")
		ping(c)
	}

	c, err = redis.DialURL("redis:///1")
	if err != nil {
		log.Printf("DB1: %v", err)
	} else {
		log.Println("DB1: got connection")
		ping(c)
	}
}
```

Expected output:

- Either both cases get the connection and succeed at pinging
- Or both fail for lack of AUTH

Actual output:

- Succeeds on DB0
- Fail for lack of AUTH on DB1

**Comment by stevenh** [comment]
Thanks for the example, I tried it here and they both fail due the auth call.
```
2021/06/08 13:22:00 DB0: got connection
2021/06/08 13:22:00 ERR Client sent AUTH, but no password is set ERR Client sent AUTH, but no password is set
2021/06/08 13:22:00 PONG <nil>
2021/06/08 13:22:00 DB1: got connection
2021/06/08 13:22:00 ERR Client sent AUTH, but no password is set ERR Client sent AUTH, but no password is set
2021/06/08 13:22:00 PONG <nil>
```
Commenting out the `log.Println(conn.Do("AUTH", pass))` results in success for both
```
2021/06/08 13:24:05 DB0: got connection
2021/06/08 13:24:05 PONG <nil>
2021/06/08 13:24:05 DB1: got connection
2021/06/08 13:24:05 PONG <nil>
```

**Comment by fgm** [comment]
That sounds like a redis server without the "requirepass" option. Is your Redis actually configured with that option and no ACL ?

**Comment by stevenh** [comment]
No it wasn't but based on your comment I was expecting conflicting behaviour between db0 and db1 but that is not the case

**Comment by stevenh** [comment]
Do note that you shouldn't be calling `AUTH` manually as that is handled by `DialURL`.

I assumed this was just used as an example, but if this is not the case then the flow you're using is wrong.

**Comment by fgm** [comment]
The thing is, when using redis://somepass@some:port/0 (instead of the complete format with the semicolon), DialURL does _not_ attempt auth on DB0, returns the connection, and the manual Do(AUTH) works. But when requesting redis://somepass@some:port/1, it does, and fails because it does not use "somepass" as the password in that case. That's the reason for that manual AUTH call.

**Comment by stevenh** [comment]
Could you test https://github.com/gomodule/redigo/pull/566 to see if that fixes the issue?

**Comment by stevenh** [comment]
> The thing is, when using redis://somepass@some:port/0 (instead of the complete format with the semicolon), DialURL does _not_ attempt auth on DB0, returns the connection, and the manual Do(AUTH) works. But when requesting redis://somepass@some:port/1, it does, and fails because it does not use "somepass" as the password in that case. That's the reason for that manual AUTH call.

To clarify this isn't a difference in behaviour of the handling of `AUTH` what you're seeing is that when a DB other than `0` is requested in the URL a `SELECT <db>` is automatically performed but as `AUTH` has not been called the request fails with a authentication error.

**Comment by fgm** [comment]
Exactly what I saw, indeed. Sorry for not being accurate enough describing the problem.

That change fixes the issue: since it applies the AUTH instead of skipping it the behaviour is now consistent, thank you.

**Comment by stevenh** [comment]
Fixed by #566
